File: python/algo/202507/algolcp05.py
from typing import List

# 常用数据结构 - 8.1 树状数组和线段树 - 树状数组
class Solution:
    def bonus(self, n: int, leadership: List[List[int]], operations: List[List[int]]) -> List[int]:
        # 参考 oi-wiki 线段树模板 最常用的求和模板
        tree =  [0] * (4 * n)   # 树组空间
        tp = [0] * (4 * n)      # 支持懒标记

        def pushdown(s, t, p):              # 区间求和
            m = s + ((t - s) >> 1)
            tree[p*2+1] += (m - s + 1) * tp[p]
            tree[p*2+2] += (t - m) * tp[p]
            tp[p*2+1] += tp[p]
            tp[p*2+2] += tp[p]
            tp[p] = 0

        def update(l:int, r:int, value:int, s:int, t:int, p:int):
            if l <= s and t <= r:
                tree[p] += (t-s+1) * value  # 本题用于区间求和 根据题意修改
                tp[p] += value
                return
            m = s + ((t - s) >> 1)
            if tp[p] and s != t:
                pushdown(s, t, p)
            if l <= m:
                update(l, r, value, s, m, p*2+1)
            if r > m:
                update(l, r, value, m + 1, t, p*2+2)
            tree[p] = tree[p*2+1] + tree[p*2+2]
        
        def query(l, r, s, t, p):
            # [l, r] 为查询区间, [s, t] 为当前节点包含的区间, p为当前节点的编号
            if l <= s and t <= r:
                return tree[p]
            if tp[p] and s != t:    # 放这里安全
                pushdown(s, t, p)
            m = s + ((t - s) >> 1)
            sum = 0
            if l <= m:
                sum += query(l, r, s, m, p * 2 + 1)
            if r > m:
                sum += query(l, r, m + 1, t, p * 2 + 2)
            return sum
        # 参考 oi-wiki 线段树模板 最常用的求和模板

        
        mod = 10 ** 9 + 7
        g = [[] for _ in range(n+1)]
        for u, v in leadership:
            g[u].append(v)
            g[v].append(u)
        
        clock, clock_in, clock_out = -1, [0] * (n+1), [0] * (n+1)
        
        def dfs(u: int, fa: int) :
            nonlocal clock
            clock += 1
            clock_in[u] = clock
            for v in g[u]:
                if v == fa: continue
                dfs(v, u)
            clock_out[u] = clock
        
        dfs(1, 0)
        ans = []
        for q in operations:
            if q[0] == 3:
                l, r = clock_in[q[1]], clock_out[q[1]]
                ans.append(query(l, r, 0, n-1, 0) % mod)
            elif q[0] == 2:
                u, w = q[1], q[2]
                update(clock_in[u], clock_out[u],w, 0, n-1, 0)
            else:
                u, w = q[1], q[2]
                update(clock_in[u], clock_in[u],w, 0, n-1, 0)

        return ans        

File: cplus/leetcode/202507/algolcp05.cpp
#include <bits/stdc++.h>
using namespace std;

// 常用数据结构 - 8.1 树状数组和线段树 - 树状数组
template<typename T>
class FenwickTree {
    vector<T> tree;

public:
    // 使用下标 1 到 n
    FenwickTree(int n) : tree(n + 1) {}
    // a[i] 增加 val, 1 <= i <= n
    void update(int i, T val) {
        for (; i < tree.size(); i += i & -i) {
            tree[i] += val;
        }
    }
    // 求前缀和 a[1] + ... + a[i] 1 <= i <= n
    T pre(int i) const {
        T res = 0;
        for (; i > 0; i &= i - 1) { // i-= i&-i
            res += tree[i];
        }
        return res;
    }
    // 求区间和 a[l] + ... + a[r], 1 <= l <= r <= n
    T query(int l, int r) const {
        if (r < l) {
            return 0;
        }
        return pre(r) - pre(l - 1);
    }
};

// 双树状数组 模板 区间更新
void range_update(FenwickTree<long>& ft1, FenwickTree<long>& ft2, int l, int r, int delta) {
    // 更新 ft1 (d[i])
    ft1.update(l, delta);
    ft1.update(r + 1, -delta);
    // 更新 ft2 (i*d[i])
    ft2.update(l, l * delta);
    ft2.update(r + 1, -(r + 1) * delta);
}

long range_query(FenwickTree<long>& ft1, FenwickTree<long>& ft2, int x) {
    return (x + 1) * ft1.pre(x) - ft2.pre(x);
}

// 双树状数组 模板 区间查询
long interval_query(FenwickTree<long>& ft1, FenwickTree<long>& ft2, int l, int r) {
    return range_query(ft1, ft2, r) - range_query(ft1, ft2, l - 1);
}

class Solution {
public:
    vector<int> bonus(int n, vector<vector<int>>& leadership, vector<vector<int>>& operations) {
        // 双树状数组，完成区间更新以及区间查询
        const int mod = 1e9 + 7;
        FenwickTree<long> ft1(n);
        FenwickTree<long> ft2(n);
        vector<vector<int>> g(n+1, vector<int>());
        for (auto q:leadership) {
            auto u = q[0], v = q[1];
            g[u].emplace_back(v);
            g[v].emplace_back(u);
        }

        vector<int> clock_in(n+1), clock_out(n+1);
        int clock = 0;
        auto dfs = [&](this auto&& dfs, int u, int fa) -> void {
            clock_in[u] = ++clock;
            for (auto v: g[u]) {
                if (v == fa) continue;
                dfs(v, u);
            }
            clock_out[u] = clock;
        };
        dfs(1, 0);

        vector<int> ans;
        for (auto q: operations) {
            if (q[0] == 3){
                ans.push_back(interval_query(ft1, ft2, clock_in[q[1]], clock_out[q[1]])%mod);
            } else if (q[0] == 2) {
                auto u = q[1], w = q[2];              
                range_update(ft1, ft2, clock_in[u], clock_out[u], w);
            } else {
                auto u = q[1], w = q[2];
                range_update(ft1, ft2, clock_in[u], clock_in[u], w);
            }
        }
        return ans;

    }
};